packages/modules/ad-personalization/src/worker/AdPersonalizationWorker.ts

```diff
@@ -54,9 +63,42 @@ export class AdPersonalizationWorker {
     await chrome.storage.local.set({ 'adPersonalization.items': adPersonalization });
   }
 
-  async send(data: SendData): Promise<MessageResponse> {
-    const { url, key } = data;
+  async getAccordantUrl(key: string, isNeedToUseNextUrl: boolean = false): Promise<string> {
+    const selectedObject = config.find((object) => object.key === key)!;
+
+    const storageData = await chrome.storage.local.get('adPersonalization.workingUrls');
+    const workingUrls = storageData['adPersonalization.workingUrls'] || {};
+
+    if (workingUrls[key] && workingUrls[key].length > 0) {
+      return workingUrls[key][0];
+    }
+
+    let currentIndex = this.urlIndexes[key] || 0;
 
+    if (isNeedToUseNextUrl) {
+      currentIndex = (currentIndex + 1) % selectedObject.url.length;
+    }
+
+    this.urlIndexes[key] = currentIndex;
+
+    return selectedObject.url[currentIndex];
+  }
+
+  private async addWorkingUrl(key: string, url: string) {
+    const storageData = await chrome.storage.local.get('adPersonalization.workingUrls');
+    const workingUrls = storageData['adPersonalization.workingUrls'] || {};
+
+    if (!workingUrls[key]) {
+      workingUrls[key] = [];
+    }
+
+    if (!workingUrls[key].includes(url)) {
+      workingUrls[key].unshift(url);
```

Comment:
Why do we need an array here? We need to store only 1 last valid URL for the user.